### DISCUSSION TITLE
need to access memory from previous job

### DISCUSSION THREAD

**Discussion by bsdis** [rq/rq#1767] [Q&A]
Hello, i need to store a variable in memory from a previous job so that i can avoid having to load data from disk on each job.
I have tried using this approach:

```py
from rq.local import LocalStack
from rq.worker import SimpleWorker
import numpy as np

_persistent = LocalStack()

def get_persistent():
    return _persistent.top


class SidoWorker(SimpleWorker):

    def work(self, *args, **kwargs):

        _persistent.push(np.ones((10,10)))
        
        #return super(Worker, self).work(*args, **kwargs)
        return super().work(*args, **kwargs)

    def execute_job(self, job: 'Job', queue: 'Queue'):
        
        _persistent.push(np.ones((10,10)))
        
        return super().execute_job(job,queue)
```
Together with:
```py
with Connection(con):
    qs = sys.argv[1:] or ['default']
    
    w = SidoWorker(qs)
    w.work()
```
Which i have seen written several places should be possible....however i just get None when i try to access `get_persistent`. I also read that `LocalStack` was not really meant for this sort of thing, so i tried different variations but nothing seems to help. 
So my question is, how do i store something in memory which i can use in other jobs such that i should not load it or send it as parameters since thats a significant overhead?

**Comment by ccrvlh** [comment]
@bsdis well, Redis is pretty much perfect for your use case, you could use something else (Python-native), but I don't really see an advantage when comparing to a proper cache datastore (Redis). You already have Redis running for the workers/queue, you could just `set/get` data, as simple as that, and in memory (fast).

**Comment by ccrvlh** [comment]
Moving this to a Q&A discussion since doesn't seem to be an issue with `RQ`, feel free to follow up on this.

**Comment by bsdis** [comment]
ok - true....but how do i obtain the redis connection inside the worker @lowercase00  ?

**Reply by ccrvlh** [reply]
You don't actually need to reuse the same connection the worker is using, or you can, I mean, doesn't really make any difference, you will just interact with your cache, normally, as you would in any other part of your code. Say you have a `app/cache.py` and then you will do something like `from app.cache import cache` and then `cache.set(...)`, there's nothing "magic" about the worker and its relation to Redis, Redis continues to be just a regular datastore.

**Reply by bsdis** [reply]
So basically just do something like
redis = Redis(
        host=settings.REDIS_HOST, 
        port=settings.REDIS_PORT, 
        db=0,
        password=settings.REDIS_PWD
    )
    
   redis.get(...) in each job?
   
   does it not have a lot of overhead?

**Reply by ccrvlh** [reply]
Yes. And like that for any datastore really. I'm not familiar with your requirements, but you would need a relevant amount of data/frequency for that to become an issue for pretty much any datastore (RDBS, File, Redis etc). Expensive is relative, but again, unless you have a fair reason to worry about that, I wouldn't worry at all.

**Reply by bsdis** [reply]
right....i am creating around 40 jobs per second for a 15 minute period (a total of almost 40K jobs) and then again for another 15 minute period later, depending on when user is recording data....
Each job takes around 20 seconds to finish... by keeping mask in memory i can reduce this time....but its also a lot of connections to create to redis....will that be an issue @lowercase00  ?

The mask is 8MB array

**Reply by ccrvlh** [reply]
What are your jobs spending time on? Network and IO or CPU?

If Network, than I would probably use Gevent/Async, if CPU, maybe threads/process, in both cases scale the workers.

That is a lot of connections for Redis, so probably using a custom worker that doesn't fork would allow you to share connections and reuse the pool.

**Reply by bsdis** [reply]
its only CPU... Heavy video processing. 
So your suggestion is to not use rq and use normal multiprocessing instead?

or is it possible to make such a custom worker using rq?

**Reply by ccrvlh** [reply]
Sure, you can make a worker that spins a new process for each new job using a pool of processes, or you can spin several processes and running a worker in each one. You want multiprocessor to avoid the GIL, and be able to run things in parallel (basically). You do want to run this with RQ, since you want to scale workers as well.

In (almost) pseudo code,  your worker could look something like:

```python
class MultiprocessWorker(Worker):
   
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.process_pool_size = kwargs.get('process_pool_size', 2)
        self.process_pool = multiprocessing.Pool(processes=self.process_pool_size)

    def execute_job(self, job, queue):
        def job_done(result):
            self.did_perform_work = True
            self.heartbeat()
            if job.get_status() == JobStatus.FINISHED:
                queue.enqueue_dependents(job)
            return result
        self.log.info("Executing job %s from %s", blue(job.id), green(queue.name))
        self.process_pool.apply_async(self.perform_job, args=(job, queue), callback=job_done)
```

So for each worker, you will start a pool of process, roughly with the number of cores on your CPU, and than you will have a custom `execute_job` method, that sends the job (that you sent to the queue) to the pool `process_pool.apply_async`, so each new job will run in a separate process. I would first experiment with the filesystem before moving to something else (you won't have the connection issue).

But again, I really don't know the details, this is just a rough idea on a possible solution to approach a CPU-bound problem. Make sure you have plenty of cores, a fast disk and a strong platform to manage the scale up / down of the workers (eg: k8s).

**Reply by bsdis** [reply]
Disk is fast (SSD). I don't have that many cores(right now), but its okay if processing keeps running for a while until all is done finishing all jobs.

I tried your example- it does not throw any errors, but it also does not really seem to do much.
I have this code:

rqworker.py:
```py
from rq.worker import Worker, JobStatus
import numpy as np
import multiprocessing

class SidoWorker(Worker):
   
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.process_pool_size = kwargs.get('process_pool_size', 2)
        self.process_pool = multiprocessing.Pool(processes=self.process_pool_size)

    def execute_job(self, job, queue):
        def job_done(result):
            self.did_perform_work = True
            self.heartbeat()
            if job.get_status() == JobStatus.FINISHED:
                queue.enqueue_dependents(job)
            print('JOB GOT DONE')
            return result
        #self.log.info("Executing job %s from %s", blue(job.id), green(queue.name))
        print("Executing job from", job.id, queue.name)
        self.process_pool.apply_async(self.perform_job, args=(job, queue), callback=job_done)
```

image_analysis.py:
```py
def process_sink_frame(filename):
    print("process_sink_frame")
    return 42
def process_float_frame(filename):
    print("process_float_frame")
    return 42
def prepare_sink_frame(filename):
    print("prepare_sink_frame")
    # job = get_current_job()
    # job.meta['handled_by'] = "me"
    # job.save_meta()
    return 42
def prepare_float_frame(filename):
    print("prepare_float_frame")
    return 42
```

workerscript.py:
```py
import sys
sys.path.append('/usr/src/app/')
from rq import Connection
from rqworker import SidoWorker
from redis import Redis
#from apps.pelletls.core.rqworker import SidoWorker

print("sys.argv", sys.argv)
#, Worker

# Preload libraries
import cv2

con = Redis(
    host='redis', 
    port=16378, 
    db=0,
    password='JfwTFQDmNGdq4jI6'
)

# Provide queue names to listen to as arguments to this script,
# similar to rq worker
with Connection(con):
    qs = sys.argv[1:] or ['default']
    
    w = SidoWorker(qs)
    w.work()
```

mainscript.py
``` py

def handleframe_success(job, connection, result):
    """worker finished successfully"""
    print('Segmentation success')

def handleframe_failure(job, connection, type, value, traceback):
    """worker failed somehow"""
    print('Segmentation fail')


                self.q.enqueue(
                    handler, 
                    args=(
                        str(filename),
                    ),
                    on_success=handleframe_success,
                    on_failure=handleframe_failure
                )

```
It seems like the worker reaches this point: `print("Executing job from", job.id, queue.name)` because thats printed to terminal....and thats the last thing i see from the jobs before they disappear in silence. No callbacks are called either. Am i missing something?

**Reply by ccrvlh** [reply]
A few things:
- Make sure you are using the right queue names: meaning you or using the `sys.argv[1:]`. I would just delete that and use the `default`. 
- Are you *sure* it's not running? I ask that because the task will run in a child process, if I'm not mistaken you won't see the `print` statements from the children (process running the task). You might need to `flush=True`, to send it to stdout
- Try doing a job that just saves a file with the PID, something like this

```python
def random_job():
    PID = current_process().pid
    f = open(f"pid_{PID}", "w")
    f.close()

    print(f"{datetime.now()} Starting job on PID {PID}")
    time.sleep(10)
    print(f"{datetime.now()} Finished job on PID {PID}")
```

then you will be able to "see" the child process running.
also  this line `self.did_perform_work = True` on the callback doesn't do anything, you can just delete it (old version).

I also think that if you kill your worker before the children finish you'll have orphaned processes, so you would probably need to the wrap the 'work()' on a try/finally block that closes() the pool and join(). You also will need to handle signals.

TBH it's probably going to be easier for you to just run a SimpleWorker in a multiprocessing pool, instead of customizing your worker. The simple worker won't fork (you already going to be an isolated process), and won't have to worry about RQ internals.

```
worker = SimpleWorker()
with Pool(4) as executor:
    pool.apply_async(worker.work)
```

With this, instead of running a job within a child process, you would be running the worker within a process, more high level, and a bit more overhead, but will be a bit easier to handle. I'm not sure whether this context manager already provides de close() and join() on __exit__, need to double check that.

EDIT: @bsdis `rq` does not have a strong support for concurrency for now. To be honest you'll be probably be better of with Celery, which already supports quite a few concurrency models. It's a bit more complicated to setup, but very powerful - worth a try in your case.

**Reply by bsdis** [reply]
hmm ok. What about just using `pool.apply_async(mytask)` directly? Wouldnt that be the easiest way to go?

**Reply by ccrvlh** [reply]
Yes. There are many reasons folks go with a task queue though. You'll get better monitoring, a lot easier to manage, easier to get status, easier to scale etc. You can't just start 40k processes

**Reply by bsdis** [reply]
right - but if i use `pool.apply_async(mytask)` aren't they then being queued up also?

**Reply by ccrvlh** [reply]
@bsdis dont get me wrong, but it's worth for you to read about that. There are a lot of trade offs and choices you'll need to make... I still recommend going to Celery and running a multiprocess worker
